File: core/src/main/java/com/loremind/application/playcontext/FrontService.java
package com.loremind.application.playcontext;

import com.loremind.domain.playcontext.Clock;
import com.loremind.domain.playcontext.Front;
import com.loremind.domain.playcontext.ports.ClockRepository;
import com.loremind.domain.playcontext.ports.FrontRepository;
import com.loremind.domain.playcontext.ports.PlaythroughRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Service applicatif des Fronts (menaces regroupant des horloges) d'une Partie.
 */
@Service
public class FrontService {

    private final FrontRepository frontRepository;
    private final ClockRepository clockRepository;
    private final PlaythroughRepository playthroughRepository;

    public FrontService(FrontRepository frontRepository, ClockRepository clockRepository,
                        PlaythroughRepository playthroughRepository) {
        this.frontRepository = frontRepository;
        this.clockRepository = clockRepository;
        this.playthroughRepository = playthroughRepository;
    }

    public List<Front> getByPlaythrough(String playthroughId) {
        return frontRepository.findByPlaythroughId(playthroughId);
    }

    public Front create(String playthroughId, String name, String description) {
        if (playthroughId == null || !playthroughRepository.existsById(playthroughId)) {
            throw new IllegalArgumentException("Partie introuvable : " + playthroughId);
        }
        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("Le nom du front est requis.");
        }
        int order = frontRepository.findByPlaythroughId(playthroughId).size();
        Front front = Front.builder()
                .playthroughId(playthroughId)
                .name(name.trim())
                .description(description)
                .order(order)
                .build();
        return frontRepository.save(front);
    }

    public Front update(String id, String name, String description) {
        Front front = require(id);
        if (name != null && !name.isBlank()) front.setName(name.trim());
        front.setDescription(description);
        return frontRepository.save(front);
    }

    /** Supprime un front et ORPHELINE ses horloges (frontId -> null : on ne perd pas d'horloges). */
    @Transactional
    public void delete(String id) {
        Front front = require(id);
        for (Clock c : clockRepository.findByPlaythroughId(front.getPlaythroughId())) {
            if (id.equals(c.getFrontId())) {
                c.setFrontId(null);
                clockRepository.save(c);
            }
        }
        frontRepository.deleteById(id);
    }

    private Front require(String id) {
        return frontRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Front introuvable : " + id));
    }
}

File: web/src/app/services/front.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Front, FrontCreate } from './front.model';

/**
 * Service HTTP pour les Fronts (menaces regroupant des horloges) d'une Partie.
 * API imbriquée sous le Playthrough : /api/playthroughs/{playthroughId}/fronts.
 */
@Injectable({ providedIn: 'root' })
export class FrontService {

  constructor(private http: HttpClient) {}

  private base(playthroughId: string): string {
    return `/api/playthroughs/${playthroughId}/fronts`;
  }

  list(playthroughId: string): Observable<Front[]> {
    return this.http.get<Front[]>(this.base(playthroughId));
  }

  create(playthroughId: string, payload: FrontCreate): Observable<Front> {
    return this.http.post<Front>(this.base(playthroughId), payload);
  }

  update(playthroughId: string, frontId: string, payload: FrontCreate): Observable<Front> {
    return this.http.put<Front>(`${this.base(playthroughId)}/${frontId}`, payload);
  }

  delete(playthroughId: string, frontId: string): Observable<void> {
    return this.http.delete<void>(`${this.base(playthroughId)}/${frontId}`);
  }
}
